src/client/datascience/jupyter/kernels/cellExecutionQueue.ts

```diff
@@ -0,0 +1,158 @@
+    }
+    /**
+     * Whether cells have been cancelled (as a result of interrupt or some have failed).
+     * Even if this property is true, its possible there is still some async operation pending (updating states of cells).
+     */
+    public get failed(): boolean {
+        return this.cancelledOrCompletedWithErrors;
+    }
+    constructor(
+        private readonly editor: NotebookEditor,
+        private readonly notebookPromise: Promise<INotebook>,
+        private readonly executionFactory: CellExecutionFactory,
+        private readonly isPythonKernelConnection: boolean
+    ) {
+        // If the editor is closed, then just stop handling the UI updates.
+        this.editor.onDidDispose(
+            async () => {
+                await this.cancel(true);
+                this.dispose();
+            },
+            this,
+            this.disposables
+        );
+    }
+    /**
+     * Queue the cell for execution & start processing it immediately.
+     */
+    public queueCell(cell: NotebookCell) {
+        const existingCellExecution = this.queueOfCellsToExecute.find((item) => item.cell === cell);
+        if (existingCellExecution) {
+            traceCellMessage(cell, 'Use existing cell execution');
+            return existingCellExecution;
+        }
+        const cellExecution = this.executionFactory.create(cell, this.isPythonKernelConnection);
+        this.queueOfCellsToExecute.push(cellExecution);
+
+        traceCellMessage(cell, 'User queued cell for execution');
+
+        // Start executing the cells.
+        this.startExecutingCells();
+    }
+    /**
+     * Cancel all cells that have been queued & wait for them to complete.
+     * @param {boolean} [forced=false]
+     * If `true`, then do not wait for cell execution to complete gracefully (just kill it).
+     * This is used when we restart the kernel (either as a result of kernel interrupt or user initiated).
+     * When restarted, the execution needs to stop as jupyter will not send more messages.
+     * Hence `forced=true` is more like a hard kill.
+     */
+    public async cancel(forced?: boolean): Promise<void> {
+        this.cancelledOrCompletedWithErrors = true;
+        traceInfo('Cancel pending cells');
+        await Promise.all(this.queueOfCellsToExecute.map((item) => item.cancel(forced)));
+    }
+    /**
+     * Wait for cells to complete (for for the queue of cells to be processed)
+     * If cells are cancelled, they are not processed, & that too counts as completion.
+     * If no cells are provided, then wait on all cells in the current queue.
+     */
+    public async waitForCompletion(cells?: NotebookCell[]): Promise<void> {
+        const cellsToCheck =
+            Array.isArray(cells) && cells.length > 0
+                ? this.queueOfCellsToExecute.filter((item) => cells.includes(item.cell))
+                : this.queueOfCellsToExecute;
+
+        await Promise.all(cellsToCheck.map((cell) => cell.result));
```

Comment:
Fixed @rchiodo let me know if this is better.
Or should i have two methods?
`waitForCompletionOfCell(cell)` & `waitForCompletionOfAllCells()`